mypy/checkexpr.py

```diff
@@ -3366,6 +3366,31 @@ def visit_op_expr(self, e: OpExpr) -> Type:
                         return proper_left_type.copy_modified(
                             items=proper_left_type.items + [UnpackType(mapped)]
                         )
+
+        if e.op == "|":
+            if is_named_instance(proper_left_type, "builtins.dict"):
+                # This is a special case for `dict | TypedDict`.
+                # Before this change this operation was not allowed due to typing limitations,
+                # however, it makes perfect sense from the runtime's point of view.
+                # So, what do we do now?
+                # 1. Find `dict | TypedDict` case
+                # 2. Switch `dict.__or__` to `TypedDict.__or__` (the same from typing's perspective)
+                # 3. Do not allow `dict.__ror__` to be executed, since this is a special case
+                # This can later be removed if `typeshed` can do this without special casing.
+                # https://github.com/python/mypy/pull/16249
```

Comment:
So do you have some specific signature(s) in mind? If not, this comment should not be here. (FWIW from what @AlexWaygood explained, it looks like the signatures are already as good as they can reasonable be.)